src/components/StateMachineProperties.tsx

```diff
@@ -0,0 +1,189 @@
+    getFirstId: (stateMachine: string) => string;
+};
+
+interface StateMachinePropertiesCallbackProps {
+    clearFinalStates: (component: string, stateMachine: string) => void;
+    setStateMachineId: (id: string) => void;
+    hideStateMachineProperties: () => void;
+    updateGraphic: (currentComponent: string, stateMachine: string) => void;
+};
+
+const mapStateToProps = (state: XCSpyState): StateMachinePropertiesProps => {
+    return {
+        active: state.stateMachineProperties.active,
+        stateMachine: state.stateMachineProperties.stateMachine,
+        id: state.stateMachineProperties.id,
+        currentComponent: state.components.currentComponent,
+        getIds: () => {
+            let componentProperties = state.components.componentProperties;
+            let currentComponent = state.components.currentComponent;
+            let stateMachine = state.stateMachineProperties.stateMachine;
+            return Object.keys(componentProperties[currentComponent].stateMachineProperties[stateMachine]);
+        },
+        getPublicMember: () => {
+            let id = state.stateMachineProperties.id;
+            if (!id)
+                return null;
+            let componentProperties = state.components.componentProperties;
+            let currentComponent = state.components.currentComponent;
+            let stateMachine = state.stateMachineProperties.stateMachine;
+            let publicMember = componentProperties[currentComponent].stateMachineProperties[stateMachine][id].jsonMessage;
+            return JSON.stringify(publicMember);
+        },
+        getStateMachineRef: () => {
+            let id = state.stateMachineProperties.id;
+            if (!id)
+                return null;
+            let componentProperties = state.components.componentProperties;
+            let currentComponent = state.components.currentComponent;
+            let stateMachine = state.stateMachineProperties.stateMachine;
+            return componentProperties[currentComponent].stateMachineProperties[stateMachine][id].stateMachineRef;
+        },
+        getFirstId: (stateMachine) => {
+            let componentProperties = state.components.componentProperties;
+            let currentComponent = state.components.currentComponent;
+            return Object.keys(componentProperties[currentComponent].stateMachineProperties[stateMachine])[0];
+        }
+
+    };
+};
+
+const mapDispatchToProps = (dispatch): StateMachinePropertiesCallbackProps => {
+    return {
+        clearFinalStates: (component, stateMachine) => {
+            dispatch(clearFinalStates(component, stateMachine));
+        },
+        setStateMachineId: (id) => {
+            dispatch(setStateMachineId(id));
+        },
+        hideStateMachineProperties: () => {
+            dispatch(hideStateMachineProperties());
+        },
+        updateGraphic: (currentComponent, stateMachine) => {
+            sessionXCSpy.getPromiseCreateSession()
```

Comment:
same comment: no xcomponent call in the ui (you have to separate ui and logic layers)